File: apps/desktop/src/lib/tauri/diagnostics.ts
import { invoke } from "@tauri-apps/api/core";

import type {
  DiagnosticEvent,
  DiagnosticEventInput,
  DiagnosticsBootstrap,
  DiagnosticSessionSummary,
  StartupRecovery,
} from "./diagnostics.types";
import { parseDiagnosticEvents, parseDiagnosticSessionSummaries } from "./diagnostics.utils";

function isTauriRuntime(): boolean {
  return typeof window !== "undefined" && "__TAURI_INTERNALS__" in window;
}

export async function bootstrapNativeDiagnostics(): Promise<DiagnosticsBootstrap | null> {
  if (!isTauriRuntime()) return null;
  return invoke<DiagnosticsBootstrap>("diagnostics_bootstrap");
}

export async function completeDiagnosticsSession(): Promise<void> {
  if (!isTauriRuntime()) return;
  await invoke("complete_diagnostics_session");
}

export async function persistDiagnosticEvent(event: DiagnosticEventInput): Promise<void> {
  if (!isTauriRuntime()) return;
  await invoke("record_diagnostic_event", { event });
}

export async function listPersistedDiagnosticSessions(): Promise<DiagnosticSessionSummary[]> {
  if (!isTauriRuntime()) return [];
  return parseDiagnosticSessionSummaries(
    await invoke<unknown>("list_persisted_diagnostic_sessions"),
  );
}

export async function readPersistedDiagnosticSessionEvents(
  sessionId: string,
): Promise<DiagnosticEvent[]> {
  if (!isTauriRuntime()) return [];
  return parseDiagnosticEvents(
    await invoke<unknown>("read_persisted_diagnostic_session_events", { sessionId }),
  );
}

export async function recordUiHeartbeat(): Promise<void> {
  if (!isTauriRuntime()) return;
  await invoke("record_ui_heartbeat");
}

export async function revealDiagnosticLogs(): Promise<void> {
  if (!isTauriRuntime()) return;
  await invoke("reveal_diagnostic_logs");
}

export async function revealDiagnosticReport(recovery: StartupRecovery): Promise<void> {
  if (!isTauriRuntime()) return;
  await invoke("reveal_diagnostic_report", { reportId: recovery.reportId });
}

File: apps/desktop/src-tauri/src/commands/diagnostics.rs
use std::sync::Arc;

use tauri::State;

use crate::{
    diagnostics::{
        DiagnosticEvent, DiagnosticEventInput, DiagnosticSessionSummary, DiagnosticsBootstrap,
        DiagnosticsState,
    },
    error::AppError,
};

#[tauri::command]
pub fn diagnostics_bootstrap(
    diagnostics: State<'_, Arc<DiagnosticsState>>,
) -> Result<DiagnosticsBootstrap, AppError> {
    diagnostics.bootstrap()
}

#[tauri::command]
pub fn complete_diagnostics_session(
    diagnostics: State<'_, Arc<DiagnosticsState>>,
) -> Result<(), AppError> {
    diagnostics.complete()
}

#[tauri::command]
pub async fn list_persisted_diagnostic_sessions(
    diagnostics: State<'_, Arc<DiagnosticsState>>,
) -> Result<Vec<DiagnosticSessionSummary>, AppError> {
    let diagnostics = Arc::clone(&diagnostics);
    tauri::async_runtime::spawn_blocking(move || diagnostics.list_persisted_sessions())
        .await
        .map_err(|_| AppError::internal("Persisted diagnostic sessions could not be listed."))?
}

#[tauri::command]
pub async fn read_persisted_diagnostic_session_events(
    diagnostics: State<'_, Arc<DiagnosticsState>>,
    session_id: String,
) -> Result<Vec<DiagnosticEvent>, AppError> {
    let diagnostics = Arc::clone(&diagnostics);
    tauri::async_runtime::spawn_blocking(move || {
        diagnostics.read_persisted_session_events(&session_id)
    })
    .await
    .map_err(|_| AppError::internal("Persisted diagnostic events could not be read."))?
}

#[tauri::command]
pub fn record_diagnostic_event(
    diagnostics: State<'_, Arc<DiagnosticsState>>,
    event: DiagnosticEventInput,
) -> Result<(), AppError> {
    diagnostics.record(event)
}

#[tauri::command]
pub fn record_ui_heartbeat(diagnostics: State<'_, Arc<DiagnosticsState>>) -> Result<(), AppError> {
    diagnostics.heartbeat()
}

#[tauri::command]
pub fn reveal_diagnostic_logs(
    diagnostics: State<'_, Arc<DiagnosticsState>>,
) -> Result<(), AppError> {
    diagnostics.reveal_logs()
}

#[tauri::command]
pub fn reveal_diagnostic_report(
    diagnostics: State<'_, Arc<DiagnosticsState>>,
    report_id: String,
) -> Result<(), AppError> {
    diagnostics.reveal_report(&report_id)
}
